core/monitor/double_sign_mointor.go

```diff
@@ -0,0 +1,149 @@
+package monitor
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/prque"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/crypto/secp256k1"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	MaxCacheHeader = 100
+)
+
+func NewDoubleSignMonitor(
+	extraSeal int,
+	sealHash func(header *types.Header) (hash common.Hash),
+) *DoubleSignMonitor {
+	return &DoubleSignMonitor{
+		sealHash:      sealHash,
+		extraSeal:     extraSeal,
+		headerNumbers: prque.New(nil),
+		headers:       make(map[uint64]*types.Header, MaxCacheHeader),
+		quit:          make(chan struct{}),
+	}
+}
+
+type DoubleSignMonitor struct {
+	extraSeal     int
+	sealHash      func(header *types.Header) (hash common.Hash)
+	headerNumbers *prque.Prque
+	headers       map[uint64]*types.Header
+	quit          chan struct{}
+}
+
+func (m *DoubleSignMonitor) getSignature(h *types.Header) ([]byte, error) {
```

Comment:
let us simplify it, because only a block header been verified (signature is right) that can go to this monitor, we dont have to recover the signer from the sender, we can trust the coinbase of the header.